contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c

```diff
@@ -245,81 +289,98 @@ tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset)
-	char		iv_prefix[16] = {0,};
-	size_t		data_size = 0;
-	XLogPageHeader curr_page_hdr = &DecryptCurrentPageHrd;
-	RelKeyData *key = NULL;
-	size_t		page_size = XLOG_BLCKSZ - offset % XLOG_BLCKSZ;
-	off_t		dec_off;
-	uint32		iv_ctr = 0;
+	ssize_t readsz;
+	char iv_prefix[16] = {0,};
+	WALKeyCacheRec *keys = pg_tde_get_wal_cache_keys();
+	XLogRecPtr write_key_lsn = 0;
+	WALKeyCacheRec *curr_key = NULL;
+	off_t dec_off = 0;
+	size_t dec_sz = 0;
+	XLogRecPtr data_start;
+	XLogRecPtr data_end;
+
+	/* !!! TODO: should be passed as the arg? */
+	static const int wal_segsz_bytes = 16 << 20; // 16Mb
 
 #ifdef TDE_XLOG_DEBUG
-	elog(DEBUG1, "read from a WAL segment, pages amount: %d, size: %lu offset: %ld", count / (Size) XLOG_BLCKSZ, count, offset);
+	elog(DEBUG1, "read from a WAL segment, size: %lu offset: %ld [%lX], seg: %X/%X", 
+					count, offset, offset, LSN_FORMAT_ARGS(segno));
 #endif
 
+	/* 
+	 * Read data from disk
+	 */
 	readsz = pg_pread(fd, buf, count, offset);
 
-	/*
-	 * Read the buf page by page and decypt ecnrypted pages. We may start or
-	 * fihish reading from/in the middle of the page (walreceiver) in such a
-	 * case we should preserve the last read page header for the IV data and
-	 * the encryption state.
-	 *
-	 * TODO: check if walsender/receiver restarts form the beggining of the
-	 * page in case of the crash.
-	 */
-	for (dec_off = 0; dec_off < readsz;)
+	if (!keys)
 	{
-		data_size = Min(page_size, readsz);
+		/* cache is empty, try to read keys from disk */
+		keys = pg_tde_fetch_wal_keys(0);
+	}
 
-		if (page_size == XLOG_BLCKSZ)
-		{
-			memcpy((char *) curr_page_hdr, (char *) buf + dec_off, SizeOfXLogShortPHD);
+#ifndef FRONTEND
+	write_key_lsn = pg_atomic_read_u64(&EncryptionState->enc_key_lsn);
+#endif
 
-			/* set the flag to "not encrypted" for the walreceiver */
-			((XLogPageHeader) ((char *) buf + dec_off))->xlp_info &= ~XLP_ENCRYPTED;
+	if (write_key_lsn != 0)
+	{
+		WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+		Assert(last_key);
 
-			Assert(curr_page_hdr->xlp_magic == XLOG_PAGE_MAGIC || curr_page_hdr->xlp_magic == 0);
-			dec_off += XLogPageHeaderSize(curr_page_hdr);
-			data_size -= XLogPageHeaderSize(curr_page_hdr);
-			/* it's a beginning of the page */
-			iv_ctr = 0;
-		}
-		else
+		/* write has generated a new key, need to fetch it */
+		if (last_key->start_lsn < write_key_lsn)
 		{
-			/* we're in the middle of the page */
-			iv_ctr = (offset % XLOG_BLCKSZ) - XLogPageHeaderSize(curr_page_hdr);
-		}
+			pg_tde_fetch_wal_keys(write_key_lsn);
 
-		if ((data_size + dec_off) > readsz)
-		{
-			data_size = readsz - dec_off;
+			/* in case cache was empty before */
+			keys = pg_tde_get_wal_cache_keys();
 		}
+	}
 
-		if (curr_page_hdr->xlp_info & XLP_ENCRYPTED)
+	SetXLogPageIVPrefix(tli, segno, iv_prefix);
+
+	XLogSegNoOffsetToRecPtr(segno, offset, wal_segsz_bytes, data_start);
+	XLogSegNoOffsetToRecPtr(segno, offset + count, wal_segsz_bytes, data_end);
+	
+	curr_key = keys;
+	while (curr_key)
```

Comment:
This while loop (and generally the use of a linked list here) seems to be far from optimal, especially for a long running server.

* We can assume that usually we want to read the more recent records - iterating from the last key would make more sense
* We can assume that multiple keys to the seq read function usually read sequentially. The key for the second call will be most likely the same key as the first key, or if not, the next one.
* And if we have many keys, using a vector and logarithmic search would be much better.

Reply:
Yeah, I know. This is one of TODOs, but I forgot to put it in writing.
Linked lists are rubbish and the implementation here is far from ideal. We shouldn't go through the key but rather pick buffer the needed range(s). But as you said, there has to be some b-tree or heap or whatever log data structure for that. There plethora of nuances that can be improved. But I'd like to make it work altogether and then improve from there.